internal/client/client.go

```diff
@@ -478,41 +572,606 @@ func (c *Client[H, Hasher, N, E, Header]) Block(hash H) (*generic.SignedBlock[N,
+					if err != nil {
+						return nil, err
+					}
+
+					stateRoot, err := operation.Op.ResetStorage(strg, stateVersion)
+					if err != nil {
+						return nil, err
+					}
+
+					if stateRoot != importHeaders.Post().StateRoot() {
+						// State root mismatch when importing state. This should not happen in
+						// safe fast sync mode, but may happen in unsafe mode.
+						logger.Warn("Error importing state: State root mismatch.")
+						return nil, blockchain.ErrInvalidStateRoot
+					}
+				}
+			}
+		}
+	}
+
+	// Ensure parent chain is finalized to maintain invariant that finality is called sequentially.
+	if finalized && parentExists && info.FinalizedHash != parentHash {
+		_, err := c.applyFinalityWithBlockHash(operation, parentHash, nil, info, makeNotifications)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var isNewBest bool
+
+	if !gapBlock && finalized {
+		switch fc := forkchoice.(type) {
+		case common.LongestChain:
+			isNewBest = importHeaders.Post().Number() > info.BestNumber
+		case common.Custom:
+			isNewBest = bool(fc)
+		default:
+			panic("unreachable")
+		}
+	}
+
+	var leafState api.NewBlockState
+
+	if finalized {
+		leafState = api.NewBlockStateFinal
+	} else if isNewBest {
+		leafState = api.NewBlockStateBest
+	} else {
+		leafState = api.NewBlockStateNormal
+	}
+
+	var treeRoute *blockchain.TreeRoute[H, N]
+
+	if isNewBest && info.BestHash != parentHash && parentExists {
+		routeFromBest, err := blockchain.NewTreeRoute(c.backend.Blockchain(), info.BestHash, parentHash)
+		if err != nil {
+			return nil, err
+		}
+		treeRoute = &routeFromBest
+	}
+
+	logger.Tracef(
+		"Imported %v, (#%d), best=%v, origin=%v",
+		hash,
+		importHeaders.Post().Number(),
+		isNewBest,
+		origin,
+	)
+
+	err = operation.Op.SetBlockData(
+		importHeaders.Post().Clone().(Header),
+		*body,
+		*indexedBody,
+		*justifications,
+		leafState,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = operation.Op.InsertAux(aux)
+	if err != nil {
+		return nil, err
+	}
+
+	c.everyImportNotificationChansMtx.Lock()
+	shouldNotifyEveryBlock := len(c.everyImportNotificationChans) > 0
+	c.everyImportNotificationChansMtx.Unlock()
+
+	// Notify when we are already synced to the tip of the chain
+	// or if this import triggers a re-org
+	shouldNotifyRecentBlock := makeNotifications || treeRoute != nil
+
+	if shouldNotifyEveryBlock || shouldNotifyRecentBlock {
+		header := importHeaders.Post()
+		if finalized && shouldNotifyRecentBlock {
+			var summary api.FinalizeSummary[H, N, Header]
+			if operation.NotifyFinalized == nil {
+				summary = api.FinalizeSummary[H, N, Header]{
+					Header:     header.Clone().(Header),
+					Finalized:  []H{parentHash},
+					StaleHeads: []H{},
+				}
+			} else {
+				operation.NotifyFinalized.Header = header.Clone().(Header)
+				operation.NotifyFinalized.Finalized = append(operation.NotifyFinalized.Finalized, parentHash)
+				summary = *operation.NotifyFinalized
+
+				operation.NotifyFinalized = nil
```

Comment:
I think this represents the `take` behaviour better.
```suggestion
				summary = *operation.NotifyFinalized
				operation.NotifyFinalized = nil
				summary.Header = header.Clone().(Header)
				summary.Finalized = append(summary.Finalized, parentHash)
```